### PR TITLE
Fix https://github.com/jeremeamia/super_closure/issues/88

### PR DESCRIPTION
**Pull request by storeman**

### CODE REVIEW

File: src/Analyzer/ClosureAnalyzer.php [hunk 1]
```diff
@@ -55,14 +55,15 @@ private function determineBinding(array &$data)
...
+		if (version_compare(PHP_VERSION, '7.0', '>=')) {
+			$closure = @$closure->bindTo(new \stdClass);
 
+			if ($closure === null) {
+				return true;
+			}
+		}
+		
         $rebound = new \ReflectionFunction($closure);
-
```
Comment (GrahamCampbell) [inline src/Analyzer/ClosureAnalyzer.php]: Please keep this newline.

**Comment by GrahamCampbell** [comment]
Thanks for the PR. Please indent with spaces though.

**Comment by storeman** [comment]
Replaced the tabs with spaces

File: src/Analyzer/ClosureAnalyzer.php [hunk 2]
```diff
@@ -55,14 +55,15 @@ private function determineBinding(array &$data)
...
-        if ($closure === null) {
-            return true;
+            if ($closure === null) {
+                return true;
+            }
         }
-
+        
         $rebound = new \ReflectionFunction($closure);
-
```
Comment (GrahamCampbell) [inline src/Analyzer/ClosureAnalyzer.php]: Please restore this newline, and remove the whitespace on the line above, then this looks good, thank you!

File: src/Analyzer/ClosureAnalyzer.php [hunk 3]
```diff
@@ -62,8 +62,8 @@ private function isClosureStatic(\Closure $closure)
                 return true;
             }
         }
-
```
Comment (GrahamCampbell) [inline src/Analyzer/ClosureAnalyzer.php]: Please keep this line, but just remove the space characters from it.

File: src/Analyzer/ClosureAnalyzer.php [hunk 4]
```diff
@@ -62,8 +62,8 @@ private function isClosureStatic(\Closure $closure)
                 return true;
             }
         }
-        
         $rebound = new \ReflectionFunction($closure);
+
```
Comment (GrahamCampbell) [inline src/Analyzer/ClosureAnalyzer.php]: Please remove the space characters, but keep the newline character.

**Comment by storeman** [comment]
Fixed it again.

**Comment by jeremeamia** [comment]
What exactly does this fix? Seems like this would cause problems, not solve them. This code was written and tested with PHP 5.

**Comment by jeremeamia** [comment]
Here is the code used in isolation. Seems to work across all supported versions.
screen shot 2016-08-26 at 6 03 48 am

**Comment by storeman** [comment]
You are right. I can't narrow the problem down to a simple set of lines. I've tried several things to reproduce it. This is the error I'm getting:

```
Cannot bind an instance to a static closure
Trace
# 	Bestand::regel 	class/methode
0 	applications/Gws3/public/index.php::19 	Phango\Application->run () 	
1 	vendor/phango/Phango/library/Phango/Application.php::212 	Phango\Http\Response->render (1) 	
2 	vendor/phango/Phango/library/Phango/Http/Response.php::55 	Phango\Http\Response->renderBody () 	
3 	vendor/phango/Phango/library/Phango/Http/Response.php::147 	Phango\Response\Body->render () 	
4 	vendor/phango/Phango/library/Phango/Response/Body.php::42 	Phango\Response\Body->renderCallers () 	
5 	vendor/phango/Phango/library/Phango/Response/Body.php::64 	Phango\Caller->call () 	
6 	vendor/phango/Phango/library/Phango/Caller.php::106 	call_user_func_array (array, array) 	
7 		PhangoAi\Controller\Index->edit (Phango\Http\PhpEnvironment\Request, 'Routeplanner', 'Route', '{"id":150073}') 	
8 	vendor/phango/PhangoAi/Controller/Index.php::386 	Phango\Db\TableAbstract->save (Routeplanner\Model\Route) 	
9 	vendor/phango/Phango/library/Phango/Db/TableAbstract.php::260 	Zend_EventManager_EventManager->trigger ('save', Routeplanner\Table\Route, array) 	
10 	vendor/phango/ZendFramework1/library/Zend/EventManager/EventManager.php::208 	Zend_EventManager_EventManager->triggerListeners ('save', Zend_EventManager_Event, ) 	
11 	vendor/phango/ZendFramework1/library/Zend/EventManager/EventManager.php::469 	call_user_func ('Routeplanner\Routeplanner::refreshRouteStatistics', Zend_EventManager_Event) 	
12 		Routeplanner\Routeplanner::refreshRouteStatistics (Zend_EventManager_Event) 	
13 	modules/Routeplanner/Routeplanner.php::26 	Phango\Table\ScheduledQueue->addItem (Closure) 	
14 	vendor/phango/Phango/Table/ScheduledQueue.php::36 	serialize (array) 	
15 		SuperClosure\SerializableClosure->serialize () 	
16 	vendor/jeremeamia/SuperClosure/src/SerializableClosure.php::111 	SuperClosure\Serializer->getData (Closure, 1) 	
17 	vendor/jeremeamia/SuperClosure/src/Serializer.php::117 	SuperClosure\Analyzer\ClosureAnalyzer->analyze (Closure) 	
18 	vendor/jeremeamia/SuperClosure/src/Analyzer/ClosureAnalyzer.php::26 	SuperClosure\Analyzer\ClosureAnalyzer->isClosureStatic (Closure) 	
19 	vendor/jeremeamia/SuperClosure/src/Analyzer/ClosureAnalyzer.php::58 	Closure->bindTo (stdClass) 	
20 		Phango\Application::exceptionErrorHandler (2, 'Cannot bind an instance to a static closure', '/var/www/checkedout.gws3/vendor/jeremeamia/SuperClosure/src/Analyzer/ClosureAnalyzer.php', 58, array) 
```
With the fix I supplied the error is resolved and everything works fine. Alson on PHP7. I'll try some other things and let you know.

**Comment by jeremeamia** [comment]
Ah ha. It turns out even with error suppression on, an error handler set with `set_error_handler()`, will still pick this up. It's up to the error handler to check this incoming `$errno` against `error_reporting()`, which apparently is 0 when the `@` error suppression operator is used, to decide whether to ignore it or not.

From [`set_error_handler()` in the PHP Manual](http://php.net/manual/en/function.set-error-handler.php):

> It is important to remember that the standard PHP error handler is completely bypassed for the error types specified by error_types unless the callback function returns FALSE. error_reporting() settings will have no effect and your error handler will be called regardless - however you are still able to read the current value of error_reporting and act appropriately. Of particular note is that this value will be 0 if the statement that caused the error was prepended by the @ error-control operator.

So, one way to fix this (in the short term) is to fix your error handler. A longer term fix for SuperClosure is to set a temporary error handler before doing `$closure->bindTo(new \stdClass);` and then restoring the previous one. Hacks either way.

**Comment by storeman** [comment]
I had a guess it had something to do with the error handler. Thanks for your support! I don't think this PR should be merged to solve this.

**Comment by jeremeamia** [comment]
Closing this PR then. Thanks.

**Comment by storeman** [comment]
I quickly tested this, but it still returns an `$errno` with a value of `2`

**Comment by jeremeamia** [comment]
Darn. Yeah, I guess we have to do my other suggestion then.

**Comment by storeman** [comment]
I found out, I shouldn't be watching `$errno` value, but the `error_reporting()` function, which will return 0 if an expression was preceded by an `@`.

**Comment by veloper** [comment]
Leaving this here incase anyone comes across this later. I ended up resolving the issue by explicitly de-binding the closure from any scope using:

```php
$newUnboundClosure = $closure->bindTo(null);
```

This worked out in my use case cause I don't want `$this` referenced in `$closure` at all.

**Pull request closed without merging**